pkg/nuclide-rust/lib/BuckIntegration.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the license found in the LICENSE file in
+ * the root directory of this source tree.
+ *
+ * @flow
+ * @format
+ */
+
+import type {TaskInfo} from '../../nuclide-buck/lib/types';
+import type {
+  AtomLanguageService,
+  LanguageService,
+} from '../../nuclide-language-service';
+
+import {getLogger} from 'log4js';
+import fsPromise from 'nuclide-commons/fsPromise';
+import {
+  getRustInputs,
+  getSaveAnalysisTargets,
+  normalizeNameForBuckQuery,
+} from './BuckUtils';
+
+import * as BuckService from '../../nuclide-buck-rpc';
+
+const logger = getLogger('nuclide-rust');
+
+export async function updateRlsBuildForTask(
+  task: TaskInfo,
+  service: AtomLanguageService<LanguageService>,
+) {
+  const buildTarget = normalizeNameForBuckQuery(task.buildTarget);
+
+  // TODO: Filter by known Rust build targets
+  const files = await getRustInputs(task.buckRoot, buildTarget);
+  // Probably not a Rust build target, ignore
+  if (files.length === 0) {
+    return;
+  }
+  // We need only to pick a representative file to get a related lang service
+  const fileUri = task.buckRoot + '/' + files[0];
+  logger.debug(`fileUri: ${fileUri}`);
+
+  const langService = await service.getLanguageServiceForUri(fileUri);
+  if (langService == null) {
+    atom.notifications.addError(`[nuclide-rust] couldn't find language service
+      for target ${buildTarget}`);
+    return;
+  }
+
+  // Since `buck` execution is not trivial - the command may be overriden, needs
+  // to inherit the environment, passes internal FB USER to env etc. the RLS
+  // can't just invoke that.
+  // Instead, we build now, copy paths to resulting .json analysis artifacts to
+  // a temp file and just use `cat $TMPFILE` as a dummy build command.
+  const analysisTargets = await getSaveAnalysisTargets(
+    task.buckRoot,
+    buildTarget,
+  );
+  logger.debug(`analysisTargets: ${analysisTargets.join('\n')}`);
+  const artifacts: Array<string> = [];
+
+  const buildReport = await BuckService.build(task.buckRoot, analysisTargets);
```

Comment:
how long does this typically take for large rust targets? do we need to report its progress?

Reply:
It depends; using `cargo check` as example, complete check build can even take a minute or two; rechecking the leaf only shouldn't take longer than a couple of seconds. Third party deps are vendored and pre-compiled from what I understand, so this should also further cut down compilation time. @jsgf would know more about how long exactly these `#save-analysis` builds take in the Buck environment.

It would certainly be good to report the progress - do you know if it's possible to pass a callback or report the progress gracefully using the Buck API here?

Reply:
Changed this to display a busy spinner indicator while the save-analysis build executes.